### ISSUE TITLE
Formatting oddities

### ISSUE THREAD

**Issue by henryiii** [tox-dev/pyproject-fmt#23]
I've been applying this to scikit-hep/cookie, and noticed some oddities in formatting. Besides #22, which is a separate issue, I'd like to dump them here instead of opening an issue for each.

---

A list of inline tables doesn't respect the indent level, that is:

```toml
authors = [
    { name = "My Name", email = "me@email.com" },
]
```

remains at four spaces.

---

`author` is placed after `maintainer`. I've generally thought author should come first, then maintainer. Fine if it's more common this way, but found it surprising.

---

```toml
dynamic = ["version"]
# turns into
dynamic = [
  "version",
]
```

Would it be possible to support Black's magic comma, and not expand a list to multiple lines if it doesn't end in a comma? Usually in black lists that are likely to change can have the trailing (magic) comma, and lists that are unlikely to change can stay short. (I thought it was supported, but just realized that `requires = ["setuptools"]` also gets spread out into multiple lines)

---

Finally, I noticed that `project.optional-dependencies` is getting sorted, which I thought wasn't supported, but it's great that it is. :)

Very happy to see this is already useful! If you let me know which ones are intended features and which ones are bugs, I can try to work on them (slowly).

**Comment by gaborbernat** [comment]
Let's not call them `oddities` please. That sounds very judgmental. A better frame of mind is just to name them questions. What is odd or not is very much a personal opinion. Formulating them as questions is much more inclusive language, and more in line with the code of conduct of the organization. Likely each of these questions would deserve its own issue to be addressed one at a time.

> A list of inline tables doesn't respect the indent level, that is:
> 
> ```toml
> authors = [
>     { name = "My Name", email = "me@email.com" },
> ]
> ```
> 
> remains at four spaces.

This is an oversight and a PR for addressing this would be welcomed.



> `author` is placed after `maintainer`. I've generally thought author should come first, then maintainer. Fine if it's more common this way, but found it surprising.

I like to put information more useful to the top. The author generally is just an artefact of who started the project back in time and carries little value to the end-user. A more important question is who is the maintainer because they are the ones that can address issues arising for the user. 

> I like dynamic at the very end, because it's special.

Currently, only classifiers follow it, so it's almost at the end. That being said I'm open to moving dynamic after that making it truly at the end.

> Would it be possible to support Black's magic comma

A PR for this would be welcome. The sole reason this doesn't work is that I did not have time to implement support for this. A similarly important feature could be adding a flag for explicit sections vs inlined sections. I personally prefer inline ones because with they can put the `dependencies` and `optional-dependencies` alongside each other, but can understand if some people would want it otherwise. Generally I'd like to keep the number of knobs low (<5), but I'm open adding a few big ones. Another flag I could accept is following the magic comma or wrap if fits in line length arg otherwise keep on one line for lists.

**Comment by henryiii** [comment]
Apologies, I didn't mean oddities in a negative way - to me it means "something unexpected" due to the way I was thinking, just like the Python oddities talk this PyCon - they mostly were not "wrong" at all, just things that might not be as one might expect.

My thinking is authors deserves a special place above maintainers, because they came first, and were the originator of the idea. And I'm saying this as someone who's more often in the maintainer list than the authors list. Also Authors is displayed above Maintainers when it is rendered on PyPI. And authors is first in https://github.com/tox-dev/tox/blob/rewrite/pyproject.toml. :D (Honestly, you might get into trouble as a maintainer putting yourself above the author if they still are around - I could see this as possibly problematic).

In the `project` table, since it's still sort of new, there's a bit more flexibility in specifying it, there aren't strong underlying conventions (unlike build-system, which has been around for years at this point). Though there should be good reasons when deviating from the PEP and documentation, as that's going to be the common basis for everyone's first project table.

FYI, I don't dislike the all-inline sections as much as I thought I might (from looking at the tox rewrite). I'm guessing the optional-dependencies might be a bit hard to write (since it will break if you have an ending comma or multiple lines in the dict, but not inside the list in the dict). The TOML authors were pushing hard to force users from writing inline tables, even going so far as to allow trailing commas and newlines in lists, but not in inline tables.

Okay, I know which ones would accept PRs now. To be clear, I'm providing an opinion and trying to convince you of it, but there will always be differing opinions, and I'm okay if you don't agree with mine and only implement yours - an unopinionated style library is way more work in providing configuration for little/no/negative benefit (see Black). :)

**Comment by gaborbernat** [comment]
> Apologies, I didn't mean oddities in a negative way - to me it means "something unexpected" due to the way I was thinking, just like the Python oddities talk this PyCon - they mostly were not "wrong" at all, just things that might not be as one might expect.

True. Things are only odd though until we understand them. So I feel in general the word question should be preferred over odd. Sounds more inclusive and less judgmental. 

> My thinking is authors deserves a special place above maintainers, because they came first, and were the originator of the idea. And I'm saying this as someone who's more often in the maintainer list than the authors list. Also Authors is displayed above Maintainers when it is rendered on PyPI. And authors is first in [https://github.com/tox-dev/tox/blob/rewrite/pyproject.toml](https://github.com/tox-dev/tox/blob/rewrite/pyproject.toml?rgh-link-date=2022-05-06T13%3A08%3A47Z). :D (Honestly, you might get into trouble as a maintainer putting yourself above the author if they still are around - I could see this as possibly problematic).

I'd personally actually prefer the other way. Even as an author once I abandon the project. Having authors show up first just means they might be contacted first for questions/issues, and once I no longer maintain a library you better not contact me anymore, because obviously, I'm no longer aware of what's the project. I really hope the Python ecosystem doesn't have too many people whose ego would require them to show up first in the file. I'd also like if PyPI would swap the order, purely because the value of maintainer in my book is always greater than authors. 


> In the `project` table, since it's still sort of new, there's a bit more flexibility in specifying it, there aren't strong underlying conventions (unlike build-system, which has been around for years at this point). Though there should be good reasons when deviating from the PEP and documentation, as that's going to be the common basis for everyone's first project table.

I never read PEPs thinking that their documentation and examples would be authoritative in the ordering/formatting. If I'd known this expectation I'd bikeshed a lot more on those.
